src/main/java/io/cryostat/core/agent/LocalProbeTemplateService.java

```diff
@@ -0,0 +1,122 @@
+ * Software and the Larger Work(s), and to sublicense the foregoing rights on
+ * either these or other terms.
+ *
+ * This license is subject to the following condition:
+ * The above copyright notice and either this complete permission notice or at
+ * a minimum a reference to the UPL must be included in all copies or
+ * substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package io.cryostat.core.agent;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Path;
+import java.util.List;
+
+import io.cryostat.core.sys.Environment;
+import io.cryostat.core.sys.FileSystem;
+
+public class LocalProbeTemplateService extends AbstractProbeTemplateService {
+
+    public static final String TEMPLATE_PATH = "CRYOSTAT_PROBE_TEMPLATE_PATH";
+
+    private final FileSystem fs;
+    private final Environment env;
+
+    public LocalProbeTemplateService(FileSystem fs, Environment env) {
+        this.fs = fs;
+        this.env = env;
+    }
+
+    public void addTemplate(InputStream inputStream) throws IOException {
+        // Sanity Check
+        if (!env.hasEnv(TEMPLATE_PATH)) {
+            throw new IOException(
+                    String.format(
+                            "Probe template directory does not exist, must be set using environment variable %s",
+                            TEMPLATE_PATH));
+        }
+        Path probeTemplateDirectory = fs.pathOf(env.getEnv(TEMPLATE_PATH));
+        // Sanity check the directory is set up correctly
+        if (!fs.exists(probeTemplateDirectory)
+                || !fs.isDirectory(probeTemplateDirectory)
+                || !fs.isReadable(probeTemplateDirectory)
+                || !fs.isWritable(probeTemplateDirectory)) {
+            throw new IOException(
+                    String.format(
+                            "Probe template directory %s does not exist, is not a directory, or has incorrect permissions.",
+                            probeTemplateDirectory.toString()));
+        }
+        try (inputStream) {
+            ProbeTemplate template = new ProbeTemplate();
+            template.deserialize(inputStream);
+            Path path = fs.pathOf(env.getEnv(TEMPLATE_PATH), template.getFileName());
+            if (fs.exists(path)) {
+                throw new Exception(
+                        String.format(
+                                "Event template \"%s\" already exists", template.getFileName()));
+            }
+            fs.writeString(path, template.serialize());
```

Comment:
The `template.deserialize` looks useful because it appears to perform some validation that the stream actually encodes a valid template, but do we need to re-serialize it to write to file? It seems like perhaps we could avoid that step if it isn't necessary, either by doing some stream piping to split it in two, or even just first copying the stream contents to file, then re-opening that new file as a stream to deserialize/validate, and then deleting the file if the validation fails.